### ISSUE TITLE
Socket leak in Python API - potential DOS

### ISSUE THREAD

**Issue by jgung** [grpc/grpc#11238]
There appears to be a socket file descriptor leak in the Python gRPC api. I was able to reproduce this following the basic tutorial here: http://www.grpc.io/docs/quickstart/python.html

Thanks in advance!

```text
ProductName:	Mac OS X
ProductVersion:	10.12.4
BuildVersion:	16E195
```
 
```text
Python 2.7.11
```
```text
appdirs==1.4.3
enum34==1.1.6
futures==3.1.1
grpcio==1.3.3
grpcio-tools==1.3.0
packaging==16.8
protobuf==3.3.0
pyparsing==2.2.0
six==1.10.0
```
 
```text
git clone https://github.com/grpc/grpc
cd grpc/examples/python/helloworld
python greeter_server.py &
nc -z localhost 50051
nc -z localhost 50051
...
lsof -n -i | grep python
```

```text
python2.7 4255 username    3u  IPv6 ...      0t0  TCP *:50051 (LISTEN)
python2.7 4255 username   11u  IPv6 ...      0t0  TCP [::1]:50051->[::1]:54211 (CLOSE_WAIT)
python2.7 4255 username   12u  IPv6 ...      0t0  TCP [::1]:50051->[::1]:54212 (CLOSE_WAIT)
```

**Comment by jgung** [comment]
Potential duplicate of #11015.

**Comment by virusdefender** [comment]
+1

I have the same issue

**Comment by liranbg** [comment]
+1 have the same issue

**Comment by nathanielmanistaatgoogle** [comment]
@jgung, @virusdefender, @liranbg: go into more detail? Is the leak here that [the sockets are only reclaimed when the garbage collector runs](https://github.com/grpc/grpc/issues/12531)? Or is this a different problem?

**Comment by nathanielmanistaatgoogle** [comment]
`<crickets>`

**Comment by jgung** [comment]
@nathanielmanistaatgoogle : thank you very much (belatedly) for the response, and apologies for the lack of response.

For context, the issue appeared for me when using a load balancer in conjunction with a gRPC server--a new socket descriptor would be created with every health check, but the resulting descriptor file would hang around. Eventually, a "too many open files" system error would occur and the server would die unexpectedly. I got around this by just disabling the health check for that particular service.

It sounds like #12531 could definitely be the issue. Testing now (grpcio=1.8.1), it appears that although the socket file descriptors do remain for a short amount of time (1 minute), they are closing in a more timely manner, so I'm happy with this issue remaining closed (though #12531 would be a nice feature).
